src/Plutarch/Extra/AssetClass.hs

```diff
@@ -1,97 +1,361 @@
+
+    -- * AssetClass - Plutarch
+    PAssetClass (PAssetClass, psymbol, pname),
+    adaSymbolData,
+    isAdaClass,
+    adaClass,
+    padaClass,
+    emptyTokenNameData,
+    pcoerceCls,
+    psymbolAssetClass,
+    pconstantCls,
+
+    -- * AssetClassData - Hask
+    AssetClassData (AssetClassData, symbol, name),
+
+    -- * AssetClassData - Plutarch
+    PAssetClassData (PAssetClassData),
+
+    -- * Scott <-> Data conversions
+    toScottEncoding,
+    ptoScottEncoding,
+    fromScottEncoding,
+    pfromScottEncoding,
+    pviaScottEncoding,
 ) where
 
+--------------------------------------------------------------------------------
+
+import GHC.TypeLits (Symbol)
+
+--------------------------------------------------------------------------------
+
+import qualified Data.Aeson as Aeson
+import Data.Tagged (Tagged, untag)
+import qualified Generics.SOP as SOP
+import qualified PlutusLedgerApi.V1.Value as Value
+
+--------------------------------------------------------------------------------
+
+import Plutarch.DataRepr (PDataFields)
+
+import Plutarch.Extra.IsData (
+    DerivePConstantViaDataList (DerivePConstantViaDataList),
+    ProductIsData (ProductIsData),
+ )
+import Plutarch.Lift (
+    PConstantDecl,
+    PUnsafeLiftDecl (PLifted),
+ )
+
+import qualified PlutusTx
+
+--------------------------------------------------------------------------------
+
+import Plutarch.Extra.Record (mkRecordConstr, (.&), (.=))
+
+import Data.Aeson (FromJSON, FromJSONKey, ToJSON, ToJSONKey)
+import qualified GHC.Generics as GHC
+
+--------------------------------------------------------------------------------
+
 import Plutarch.Api.V1 (
     PCurrencySymbol,
     PTokenName,
-    PValue (PValue),
- )
-import Plutarch.Api.V1.AssocMap (PMap (PMap))
-import Plutarch.Api.V2 (
-    AmountGuarantees,
-    KeyGuarantees,
  )
-import Plutarch.DataRepr (PDataFields)
-import Plutarch.Extra.TermCont (pletC, pmatchC)
+import Plutarch.Unsafe (punsafeCoerce)
+
+--------------------------------------------------------------------------------
+
+import Plutarch.Orphans ()
+
+--------------------------------------------------------------------------------
+
+import qualified PlutusLedgerApi.V1.Value as Plutus
+
+--------------------------------------------------------------------------------
+-- AssetClass & Variants
+
+{- | Version of Plutus AssetClass, with currency unit tag
+ @since 3.8.0
+-}
+data AssetClass (tag :: Symbol) = AssetClass
+    { symbol :: Plutus.CurrencySymbol
+    , name :: Plutus.TokenName
+    }
+    deriving stock
+        ( -- | @since 3.8.0
+          Eq
+        , -- | @since 3.8.0
+          Ord
+        , -- | @since 3.8.0
+          Show
+        , -- | @since 3.8.0
+          GHC.Generic
+        )
+    deriving anyclass
+        ( -- | @since 3.8.0
+          ToJSON
+        , -- | @since 3.8.0
+          FromJSON
+        , -- | @since 3.8.0
+          FromJSONKey
+        , -- | @since 3.8.0
+          ToJSONKey
+        )
```

Comment:
Did any of these instances exist on the old `AssetClass`? If so, they shouldn't be tagged as `since` 3.8.0. Only new instance declarations should be.

Reply:
The old asset class was untagged, so I'm assuming this would make a difference?

Reply:
What do you think?